### DISCUSSION TITLE
Adding fake cursors without secondary selection

### DISCUSSION THREAD

**Discussion by Esnos33** [meow-edit/meow#669] [Ideas]
Hi, my problem with beacon mode is that it is slow to deploy, because you first have to select the whole text where you want to look for things, then you have to make a secondary selection and then choose the things you want. I think it would be much faster if you could just add a fake-cursor at a point whenever you like and when you record kmacro, you would just run this kmacro once at each point with a fake-cursor. For example, I would like to meow-word, add fake cursor with meow-add-fake-cursor or something like this, meow-search, meow-add-fake-cursor, ... , record kmacro -> kmacro would start at each of fake-cursors.

**Comment by DogLooksGood** [comment]
You don't have to use beacon for kmacros. if you want to apply the same operation to some positions, you should record the kmacro at the first position, then move your cursor to next position, call the kmacro, then repeat. There's no need to add cursors manually before you record the operation, it doesn't benefit at all.

**Comment by Esnos33** [comment]
My main point is that searching and adding fake cursors for all points of interest in one go and then once calling kmacro takes fewer keypresses and overhead than repeatedly calling kmacro and searching for just one point of interest at a time. I think the main reason people like multiple-cursors is not because of live preview, but because they can divide task into two big steps instead of multiple small ones and this feature would give them this ability. The main benefit of this feature over multiple-cursors would be performance because emacs uses a gap buffer and having all kmacro build-in support.

**Reply by DogLooksGood** [reply]
I think it doesn't take fewer keys as calling a command is required to set a beacon, or maybe more commands for setting beacons with different directions and types. And obviously there's no enough key spaces for binding these commands to short key sequences.

**Reply by Esnos33** [reply]
You save key presses because you don't have to activate search every time and you don't have to think about not populating regexp-search-ring during kmacro. About setting beacons with different directions and types, it could work in the same way as it works now with cursors in beacon mode, with the difference that region would be unbounded and you would have only two new commands, meow-mark-next-like-this and meow-mark-all-like-this, which would populate fake cursors based on last command in normal mode. 
For example in this starting buffer where | is point, * is mark, ^ is fake point and $ is fake mark:
```
banan-mango apple kiwi
*|apple banan-mango kiwi
kiwi apple banan-mango 
```
calling meow-mark-next-like-this would add ^ under *|, like in beacon mode
```
banan-mango apple kiwi
*|apple banan-mango kiwi
$^kiwi apple banan-mango 
```
and recording kmacro like `i fruit<spc> esc F3 F4` would run this kmacro at each fake cursor
```
banan-mango apple kiwi
fruit0 *|apple banan-mango kiwi
fruit1 $^kiwi apple banan-mango 
```

Next example, with same starting buffer
```
banan-mango apple kiwi
*|apple banan-mango kiwi
kiwi apple banan-mango 
```
call meow-mark-word to select and have it on in regexp-search-ring
```
banan-mango apple kiwi
*apple| banan-mango kiwi
kiwi apple banan-mango 
```
calling meow-mark-next-like-this would add ^ after next apple, like in beacon mode
```
banan-mango apple kiwi
*apple| banan-mango kiwi
kiwi $apple^ banan-mango 
```
After calling same kmacro as before (`i fruit<spc> esc F3 F4`)
```
banan-mango apple kiwi
fruit0 *|apple banan-mango kiwi
kiwi fruit1 $^apple banan-mango 
```

Last example with same starting buffer
```
banan-mango apple kiwi
*|apple banan-mango kiwi
kiwi apple banan-mango 
```
call meow-find - to move to -
```
banan-mango apple kiwi
*apple banan-|mango kiwi
kiwi apple banan-mango 
```
calling meow-mark-all-like-this would add ^ after each - in buffer and move original mark to point, like in beacon mode
```
banan-$^mango apple kiwi
apple banan-*|mango kiwi
kiwi apple banan-$^mango 
```
After calling same kmacro as before (`i fruit<spc> esc F3 F4`)
```
banan-fruit2 $^mango apple kiwi
apple banan-fruit0 *|mango kiwi
kiwi apple banan-fruit1 $^mango 
```

**Reply by DogLooksGood** [reply]
I think in current implementation, what you should do for such modification is
(Let's take the second case as example)

```
banan-mango apple kiwi
*|apple banan-mango kiwi
kiwi apple banan-mango 
```

Call `meow-line` 2 times. Then we have

```
banan-mango apple kiwi
*apple banan-mango kiwi
kiwi apple banan-mango| 
```

Call `meow-grab` and `meow-pop-to-mark`, then we have (`[]` stands for secondary selection)

```
banan-mango apple kiwi
[*|apple banan-mango kiwi
kiwi apple banan-mango] 
```

Call `meow-mark-word` then we have

```
banan-mango apple kiwi
[*apple| banan-mango kiwi
kiwi $apple^ banan-mango] 
```

Then we can edit. 

Yes it takes more keys, in this case it's 3 more keys. But I don't think it's a big deal, as you have all the methods to define the range. If we create the fake cursor first, we will be in a middle state, the only option we have is to create one more cursor(or all possible cursors in buffer). It's not really flexible. There's no easy way to express something like, "now create all cursors like this in this `{...}` block" without introducing more commands.

**Comment by eeriemyxi** [comment]
I had the same thoughts. My solution is to bind `V` to `call-last-kbd-macro`. Select what you want search, then do `v;v`, now do F3 and start doing what you want to do with other cursors. Once done do F4 then press `v` to go next match and just do `V` to apply the changes. If you want to ignore a match, just don't do `V` and press `v` instead. If you want to go to the previous match then do `;v`. This should also work just fine for `/`, `m`, and `M`.

Alternatives to F3 and F4 include `C-x (` and `C-x )`.
